### ISSUE TITLE
[Optimization] Fuse projections (QKV + u + gate) into one GEMM

### ISSUE THREAD

**Issue by z-a-f** [maida-ai/Research#44]
### Summary
Replace 5 separate Linear ops (`W_Q, W_K, W_V, B, gate`) with one `nn.Linear` and `torch.split`.

### Motivation
Cuts kernel launches and DRAM traffic; generally +10-30% step speed.

### Scope

* Add `self.W_fused = nn.Linear(d_model, 3*d_model + ssm_state_dim + d_model)`.
* In forward paths, do:

  * `fused = self.W_fused(x_norm1)`
  * `Q, K, V, u, gate_pre = torch.split(fused, [d,d,d,d_s,d], dim=-1)`
  * Use `u` in SSM instead of `self.B(x)`.
  * `g = torch.sigmoid(gate_pre)`.

### Implementation Sketch

```python
# __init__: remove W_Q, W_K, W_V, B, gate; add W_fused
self.W_fused = nn.Linear(d_model, 3*d_model + self.ssm_state_dim + d_model)

# forward:
fused = self.W_fused(x_norm1)
Q, K, V, u, gate_pre = torch.split(fused, [d, d, d, self.ssm_state_dim, d], dim=-1)
g = torch.sigmoid(gate_pre)
# reshape Q,K,V to heads; pass u to _compute_ssm via arg so it skips B(x)
```

### Tests

* Numerical parity (within tolerance) vs. previous version on random input.
* Parameter count change is expected but reasonable.
* Throughput increases on T∈{512,1024}.

### Acceptance Criteria

* [ ] All tests pass; measurable speedup.
* [ ] Code paths updated to use `u` instead of `self.B`.

### Rollout notes

* Keep new behavior **behind flags** where practical (`ssm_impl`, `tile_size`, `num_kv_heads`, `proj_stride`, `overlap_streams`, `max_mask_T`, `compile`).
* Add quick **unit tests** (shape, state continuity, streaming equivalence).
* In each PR description, include a **tiny benchmark snippet** (tokens/sec @ T=512 or 1024) to document gains.

### Context

Replace `W_Q/W_K/W_V/B/gate` with a single `nn.Linear(d, 3d + d_s + d)` and `torch.split`. Fewer kernel launches & less DRAM traffic → typical **+10-30%** step speed.

```python
# __init__
self.W_fused = nn.Linear(d_model, 3*d_model + self.ssm_state_dim + d_model, bias=True)

# forward
fused = self.W_fused(x_norm1)  # (B,T, 3d + d_s + d)
Q, K, V, u, gate_pre = torch.split(fused, [d, d, d, self.ssm_state_dim, d], dim=-1)
g = torch.sigmoid(gate_pre)
# use u directly in SSM (drop self.B)
```
